packages/graph/hash_graph/lib/graph/src/store/query/mod.rs

```diff
@@ -1,10 +1,259 @@
+            false
+        }
+    }
+}
+
+impl From<serde_json::Value> for Literal {
+    fn from(value: serde_json::Value) -> Self {
+        use serde_json::Value;
+
+        match value {
+            Value::Null => Self::Null,
+            Value::Bool(bool) => Self::Bool(bool),
+            // TODO: Support Integer to avoid potential panic
+            //   see https://app.asana.com/0/0/1202884883200973/f
+            Value::Number(number) => number.as_f64().map_or_else(
+                || panic!("Could not parse {number} as literal"),
+                Self::Float,
+            ),
+            Value::String(string) => Self::String(string),
+            Value::Array(list) => Self::List(list.into_iter().map(From::from).collect()),
+            Value::Object(_) => {
+                // see: https://app.asana.com/0/0/1202884883200943/f
+                todo!("`Literal::Object`")
+            }
+        }
+    }
+}
+
+type Identifier = String;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(transparent)]
+pub struct PathSegment {
+    pub identifier: Identifier,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(transparent)]
+pub struct Path {
+    pub segments: Vec<PathSegment>,
+}
+
+// TODO: DOC: Write documentation for the AST
+//   see https://app.asana.com/0/0/1202884883200976/f
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub enum Expression {
+    Eq(Vec<Expression>),
+    Ne(Vec<Expression>),
+    All(Vec<Expression>),
+    Any(Vec<Expression>),
+    Literal(Literal),
+    Path(Path),
+    Field(Identifier),
+}
+
+impl Expression {
+    #[must_use]
+    pub fn for_versioned_uri(uri: &VersionedUri) -> Self {
+        Self::All(vec![
+            Self::Eq(vec![
+                Self::Path(Path {
+                    segments: vec![PathSegment {
+                        identifier: "version".to_owned(),
+                    }],
+                }),
+                Self::Literal(Literal::Float(f64::from(uri.version()))),
+            ]),
+            Self::Eq(vec![
+                Self::Path(Path {
+                    segments: vec![PathSegment {
+                        identifier: "uri".to_owned(),
+                    }],
+                }),
+                Self::Literal(Literal::String(uri.base_uri().to_string())),
+            ]),
+        ])
+    }
+
+    #[must_use]
+    pub fn for_latest_version() -> Self {
+        Self::Eq(vec![
+            Self::Path(Path {
+                segments: vec![PathSegment {
+                    identifier: "version".to_owned(),
+                }],
+            }),
+            Self::Literal(Literal::String("latest".to_owned())),
+        ])
+    }
+}
+
+impl Expression {
+    // TODO: Implement error handling
+    //   see https://app.asana.com/0/0/1202884883200968/f
+    #[allow(clippy::missing_panics_doc, reason = "Error handling not applied yet")]
+    pub fn evaluate<'t, 'resolver, 'context, 'r, R, C>(
```

Comment:
Why do we need `'r`, it doesn't seem to be used.

I'm also not a massive fan of the choice of lifetimes here, I think it goes against our recommendations:
> We decided on lifetimes:
> - Avoid `'a`, `'b`, `'c`
> - Try to match the actual reference, like `&'r Report` or `&'p Provider`
> - If single letter lifetimes are confusing in a context, try to abbreviate it: `&'rep Report` or `&'req Request`

(We have yet to persist this, I've set a reminder for myself)

Reply:
I forgot to do cleanup here. It was hard to figure out all the different lifetimes, so having longer lifetime names was easier. Will clean them up 🙂

Reply:
No worries :) Give me a ping when you want me to take another look